ThunderCloud/TSCAppDelegate.swift

```diff
@@ -0,0 +1,196 @@
+		if window?.rootViewController?.presentedViewController != nil {
+			
+			DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+				_ = self.handleNotificationResponse(for: notification, response: response, fromLaunch: fromLaunch)
+			}
+			
+		} else {
+			
+			return handleContentNotificationFor(cacheURL: url)
+		}
+		
+		return true
+	}
+	
+	open func handleContentNotificationFor(cacheURL: String) -> Bool {
+		
+		if Bundle.main.infoDictionary?["TSCStreamingBaseURL"] as? String != nil {
+			
+			// Stream the page
+			if let window = window {
+				MDCHUDActivityView.start(in: window)
+			}
+			
+			let streamingController = StreamingPagesController()
+			streamingController.fetchStreamingPage(cacheURLString: cacheURL, completion: { (stormViewController, error) in
+				
+				OperationQueue.main.addOperation {
+					
+					guard let window = self.window else { return }
+					
+					MDCHUDActivityView.finish(in: window)
+					
+					if let error = error {
+						UIAlertController.presentError(error, in: window.rootViewController!)
+					}
+					
+					guard let viewController = stormViewController else { return }
+					
+					viewController.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(self.dismissStreamedPage))
+					let navController = UINavigationController(rootViewController: viewController)
+					window.rootViewController?.present(navController, animated: true, completion: nil)
+				}
+			})
+			
+			return true
+			
+		} else {
+			
+			// Load the page locally
+			guard let pageURL = URL(string: cacheURL) else {
+				return false
+			}
+			
+			guard let viewController = TSCStormViewController(url: pageURL) else { return false }
+			
+			viewController.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: viewController, action: #selector(UIViewController.dismissAnimated))
+			let navController = UINavigationController(rootViewController: viewController)
+			window?.rootViewController?.present(navController, animated: true, completion: nil)
+			
+			return true
+		}
+	}
+	
+	public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+		
+		let application = UIApplication.shared
+		_ = handleNotificationResponse(for: response.notification, response: response, fromLaunch: application.applicationState == .inactive || application.applicationState == .background)
+		
+		completionHandler()
+	}
+	
+	open func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
+		
+		if let foregroundOptions = foregroundNotificationOptions {
+			completionHandler(foregroundOptions)
+			return
+		}
+		
+	    _ = handleNotificationResponse(for: notification, response: nil, fromLaunch: false)
```

Comment:
Could this be shifted along to be in line with the rest of the function please? :)